File: BEmu/IntradayTickRequest/ElementIntradayTickResponseError.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="IntradayTickRequest/ElementIntradayTickSecurityError.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.IntradayTickRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementIntradayTickResponseError : Element
    {
        private readonly ElementIntradayTickString _source, _category, _message, _subCategory;
        private readonly ElementIntradayTickInt _code;

        internal ElementIntradayTickResponseError()
        {
            int code = Types.RandomDataGenerator.RandomInt(99);
            string sourceGibberish = Types.RandomDataGenerator.RandomString(5).ToLower();

            this._source = new ElementIntradayTickString("source", string.Format("{0}::{1}{2}", code, sourceGibberish, Types.RandomDataGenerator.RandomInt(99)));
            this._code = new ElementIntradayTickInt("code", code);
            this._category = new ElementIntradayTickString("category", "BAD_SEC");
            this._message = new ElementIntradayTickString("message", string.Format("Unknown/Invalid security [nid:{0}]", code));
            this._subCategory = new ElementIntradayTickString("subcategory", "INVALID_SECURITY");
        }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._source;
                yield return this._code;
                yield return this._category;
                yield return this._message;
                yield return this._subCategory;
            }
        }

        public override Name Name { get { return new Name("responseError"); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return 5; } }

        public override string GetElementAsString(string name)
        {
            return this[name].GetValueAsString();
        }

        public override int GetElementAsInt32(string name)
        {
            return this[name].GetValueAsInt32();
        }

        public override Element GetElement(string name)
        {
            return this[name];
        }

        public override object GetValue()
        {
            return null;
        }

        public override object GetValue(int index)
        {
            return null;
        }

        public override bool IsArray
        {
            get
            {
                return false;
            }
        }

        public override bool IsComplexType
        {
            get
            {
                return true;
            }
        }

        public override bool IsNull
        {
            get
            {
                return false;
            }
        }

        public override object this[int index]
        {
            get
            {
                return null;
            }
        }

        public override Element this[string name]
        {
            get
            {
                string strName = name.ToUpper();
                foreach (var item in this.Elements)
                {
                    if (item.Name.ToString().ToUpper() == strName)
                        return item;
                }
                return base[name];
            }
        }

        public override object this[string name, int index]
        {
            get
            {
                if (index == 0)
                    return this[name];
                else
                    throw new IndexOutOfRangeException();
            }
        }

        public override bool HasElement(string name, bool excludeNullElements = false)
        {
            foreach (var item in this.Elements)
            {
                if (item.Name.ToString().ToUpper() == name.ToUpper())
                    return true;
            }
            return false;
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}responseError = {{{1}", tabs, Environment.NewLine);
            foreach (var item in this.Elements)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);
            return result;
        }

    }
}

File: BEmu_cpp/src/IntradayTickRequest/ElementIntradayTickResponseError.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/IntradayTickRequest/ElementIntradayTickResponseError.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "IntradayTickRequest/ElementIntradayTickResponseError.h"
#include "BloombergTypes/Name.h"
#include "BloombergTypes/ElementPtr.h"
#include "Types/RandomDataGenerator.h"
#include "IntradayTickRequest/ElementIntradayTickString.h"
#include "IntradayTickRequest/ElementIntradayTickInt.h"
#include "Types/IndentType.h"

namespace BEmu
{
	namespace IntradayTickRequest
	{
		ElementIntradayTickResponseError::ElementIntradayTickResponseError()
		{
			int code = RandomDataGenerator::RandomInt(99);
			std::string* sourceGibberish = RandomDataGenerator::RandomString(5);
			std::transform(sourceGibberish->begin(), sourceGibberish->end(), sourceGibberish->begin(), ::tolower);

			//source
			std::stringstream sourceValue;
			sourceValue << code << "::" << sourceGibberish << RandomDataGenerator::RandomInt(99);
			this->_source = new ElementIntradayTickString("source", sourceValue.str());
			delete sourceGibberish;
			sourceGibberish = 0;

			//code
			this->_code = new ElementIntradayTickInt("code", code);

			//category
			this->_category = new ElementIntradayTickString("category", "BAD_SEC");

			//message
			std::stringstream messageValue;
			messageValue << "Unknown/Invalid security [nid:" << code << "]";
			this->_message = new ElementIntradayTickString("message", messageValue.str());

			//sub-category
			this->_subCategory = new ElementIntradayTickString("subcategory", "INVALID_SECURITY");
		}

		ElementIntradayTickResponseError::~ElementIntradayTickResponseError()
		{
			delete this->_source;
			this->_source = 0;

			delete this->_code;
			this->_code = 0;

			delete this->_category;
			this->_category = 0;

			delete this->_message;
			this->_message = 0;

			delete this->_subCategory;
			this->_subCategory = 0;
		}

		Name ElementIntradayTickResponseError::name() const
		{
			Name result("responseError");
			return result;
		}

		const char* ElementIntradayTickResponseError::getElementAsString(const char* name) const
		{
			return this->getElement(name)->getValueAsString(0);
		}

		int ElementIntradayTickResponseError::getElementAsInt32(const char* name) const
		{
			return this->getElement(name)->getValueAsInt32(0);
		}

		ElementPtr * ElementIntradayTickResponseError::getElement(const char* name) const
		{
			if(strncmp("source", name, 6) == 0)
				return this->_source;

			else if(strncmp("code", name, 4) == 0)
				return this->_code;

			else if(strncmp("category", name, 8) == 0)
				return this->_category;

			else if(strncmp("message", name, 7) == 0)
				return this->_message;

			else if(strncmp("subcategory", name, 11) == 0)
				return this->_subCategory;

			else
				throw elementPtrEx;
		}

		bool ElementIntradayTickResponseError::hasElement(const char* name, bool excludeNullElements) const
		{
			return
				(strncmp("source", name, 6) == 0) ||
				(strncmp("code", name, 4) == 0) ||
				(strncmp("category", name, 8) == 0) ||
				(strncmp("message", name, 7) == 0) ||
				(strncmp("subcategory", name, 11) == 0);
		}

		std::ostream& ElementIntradayTickResponseError::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs = IndentType::Indent(level, spacesPerLevel);

			stream << tabs << "responseError = {" << std::endl;
			this->_source->print(stream, level + 1, spacesPerLevel);
			this->_code->print(stream, level + 1, spacesPerLevel);
			this->_category->print(stream, level + 1, spacesPerLevel);
			this->_message->print(stream, level + 1, spacesPerLevel);
			this->_subCategory->print(stream, level + 1, spacesPerLevel);
			stream << tabs << '}' << std::endl;

			return stream;
		}

	}
}